ddtrace/contrib/cherrypy/middleware.py

```diff
@@ -0,0 +1,139 @@
+            span_type=SpanTypes.WEB,
+        )
+
+    def _after_error_response(self):
+        span = getattr(cherrypy.request, "_datadog_span", None)
+
+        if not span:
+            log.warning("cherrypy: tracing tool after_error_response hook called, but no active span found")
+            return
+
+        if not span.sampled:
+            return
+
+        span.error = 1
+        span.set_tag(errors.ERROR_TYPE, cherrypy._cperror._exc_info()[0])
+        span.set_tag(errors.ERROR_MSG, str(cherrypy._cperror._exc_info()[1]))
+        span.set_tag(errors.STACK, cherrypy._cperror.format_exc())
+
+        self._close_span(span)
+
+    def _on_end_request(self):
+        span = getattr(cherrypy.request, "_datadog_span", None)
+
+        if not span:
+            log.warning("cherrypy: tracing tool on_end_request hook called, but no active span found")
+            return
+
+        if not span.sampled:
+            return
+
+        self._close_span(span)
+
+    def _close_span(self, span):
+        # Let users specify their own resource in middleware if they so desire.
+        # See case https://github.com/DataDog/dd-trace-py/issues/353
+        if span.resource == SPAN_NAME:
+            # In the future, mask virtual path components in a
+            # URL e.g. /dispatch/abc123 becomes /dispatch/{{test_value}}/
+            # Following investigation, this should be possible using
+            # [find_handler](https://docs.cherrypy.org/en/latest/_modules/cherrypy/_cpdispatch.html#Dispatcher.find_handler)
+            # but this may not be as easy as `cherrypy.request.dispatch.find_handler(cherrypy.request.path_info)` as
+            # this function only ever seems to return an empty list for the virtual path components.
+
+            # For now, default resource is method and path:
+            #   GET /
+            #   POST /save
+            resource = "{} {}".format(cherrypy.request.method, cherrypy.request.path_info)
+            span.resource = compat.to_unicode(resource)
+
+        url = compat.to_unicode(cherrypy.request.base + cherrypy.request.path_info)
+        status_code, _, _ = valid_status(cherrypy.response.status)
+
+        trace_utils.set_http_meta(
+            span,
+            config.cherrypy,
+            method=cherrypy.request.method,
+            url=url,
+            status_code=status_code,
+            request_headers=cherrypy.request.headers,
+            response_headers=cherrypy.response.headers,
+        )
+
+        span.finish()
+
+        # Clear our span just in case.
+        cherrypy.request._datadog_span = None
+
+
+class TraceMiddleware(object):
+    def __init__(self, app, tracer, service="cherrypy", use_signals=True, distributed_tracing=True):
+        self.app = app
+
+        # save our traces.
+        self._tracer = tracer
+        self._service = service
+        self._use_distributed_tracing = distributed_tracing
+        self.use_signals = use_signals
```

Comment:
Is this actually necessary? I'm not sure if this is a cherrypy thing that's required or not but just wanted to make sure as I don't see it used except in a test case 🙂

Reply:
No, looks like it has been there since 2017 and never used. I presume a copy and paste from old. Removed.